crates/floresta-wire/src/p2p_wire/tests/chain_selector.rs

```diff
@@ -0,0 +1,537 @@
+                assert_eq!(
+                    chain.get_root_hashes()[3],
+                    NodeHash::from_str(
+                        "bfe030a7a994b921fb2329ff085bd0f2351cb5fa251985d6646aaf57954b782b"
+                    )
+                    .unwrap()
+                );
+                assert_eq!(chain.get_root_hashes().len(), 6);
+                assert_eq!(chain.get_best_block().unwrap().1, headers[119].block_hash());
+            }
+            "ten_peers_one_honest" => {
+                assert_eq!(chain.is_in_idb(), false);
+                assert_eq!(
+                    chain.get_root_hashes()[3],
+                    NodeHash::from_str(
+                        "bfe030a7a994b921fb2329ff085bd0f2351cb5fa251985d6646aaf57954b782b"
+                    )
+                    .unwrap()
+                );
+                assert_eq!(chain.get_root_hashes().len(), 6);
+                assert_eq!(chain.get_best_block().unwrap().1, headers[119].block_hash());
+            }
+            _ => {}
+        }
+    }
+
+    pub fn get_essentials() -> (
+        Vec<Header>,
+        HashMap<BlockHash, UtreexoBlock>,
+        HashMap<BlockHash, Vec<u8>>,
+        BlockHash,
+    ) {
+        let headers = get_test_headers();
+        let blocks = get_test_blocks().unwrap();
+        let true_filters = get_test_filters("./src/p2p_wire/tests/test_data/roots.json").unwrap();
+
+        // // BlockHash of chain_tip: 0000035f0e5513b26bba7cead874fdf06241a934e4bc4cf7a0381c60e4cdd2bb (119)
+        let tip_hash =
+            BlockHash::from_str("0000035f0e5513b26bba7cead874fdf06241a934e4bc4cf7a0381c60e4cdd2bb")
+                .unwrap();
+
+        (headers, blocks, true_filters, tip_hash)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use super::tests_utils::create_false_acc;
+    use super::tests_utils::get_essentials;
+    use super::tests_utils::setup_test;
+
+    #[async_std::test]
+    async fn accept_one_header() {
+        let (headers, _, _, _) = get_essentials();
+
+        setup_test(
+            "test_chain_selector",
+            vec![(headers, HashMap::new(), HashMap::new())],
+            false,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn two_peers_different_tips() {
+        let (mut headers, _, _, _) = get_essentials();
+
+        let mut peers = Vec::new();
+
+        for _ in 0..2 {
+            headers.pop();
+            peers.push((headers.clone(), HashMap::new(), HashMap::new()))
+        }
+
+        setup_test(
+            "two_peers_different_tips",
+            peers,
+            false,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn ten_peers_different_tips() {
+        let (mut headers, _, _, _) = get_essentials();
+
+        let mut peers = Vec::new();
+
+        for _ in 0..10 {
```

Comment:
In some cases where the tip is different, could you please make them an actual fork? Like:

```
1 -> 2 -> 3
  \-> 2a
```

on regtest you can easily create those with `invalidateblock`. Just mine a couple of blocks, take the headers and roots. Then you invalidate some in the middle and mine some more